### PR TITLE
python: Restore logging whether or not the dependency has been found

### PR DESCRIPTION
**Pull request by xclaesse**

### CODE REVIEW

**Comment by dcbaker** [comment]
I'm happy with patch 1 r-b on that, patch two not so much. We already have the type information in the form of the class, and i really don't like that python dependencies suddenly have different behavior than every other kind of dependency

**Comment by xclaesse** [comment]
> We already have the type information in the form of the class

That's true, changed to have method string as class member instead of passing to contructor everywhere.

> i really don't like that python dependencies suddenly have different behavior than every other kind of dependency

I think it's useful information to know the method that found python dependency. This just restore the previous behavior that was lost, it's not new.

**Comment by dcbaker** [comment]
But it's still specific to the python dependency, that should be handled for all dependencies in a single centralized place, not one off for just python.

**Comment by xclaesse** [comment]
@dcbaker we should use factory infrastructure indeed, but that's a bigger refactoring that should not go into 0.60.1. We can always rethink this for 0.61.

**Comment by dcbaker** [comment]
Then patch two shouidn't go into 0.60.1 either, since it's not a bug fix?

**Comment by xclaesse** [comment]
@dcbaker it is a regression, it's not reported in the logs any more if it found python dep or not. Can we stop useless bikeshedding here please?

**Comment by nirbheek** [comment]
I think Dylan's concern is valid, so it sounds to me like we should:

1. Split patch 2 out of this PR
2. Merge this PR immediately and release 0.60.1
3. Do a proper fix for logging what method found the python dependency, and merge it into master
4. Open a PR targeting the 0.60 branch with patch 2, and merge it into 0.60.2

Does this sound ok?

**Comment by xclaesse** [comment]
That sounds like an awfully complicated plan for something that could have been merged already instead of discussing non-issues. Do as you wish, I don't care.

**Comment by eli-schwartz** [comment]
Since everyone seems to agree that the first patch is good and should be merged ASAP and added to 0.60.1, I merged it manually as c8d125653d0535624741c591397a2b9b4ae3bb43.

We can now proceed to discuss whether or not adding more info back to the log message is important enough to merge a disputed change for urgent backporting into 0.60.1 -- without holding up the first, undisputed, patch.

**Comment by nirbheek** [comment]
> Since everyone seems to agree that the first patch is good and should be merged ASAP and added to 0.60.1, I merged it manually as [c8d1256](https://github.com/mesonbuild/meson/commit/c8d125653d0535624741c591397a2b9b4ae3bb43).

I've cherry-picked this into 0.60 and moved the milestone on this PR to 0.60.2

File: test cases/unit/101 python without pkgconfig/meson.build
Comment (Conan-Kudo) [inline test cases/unit/101 python without pkgconfig/meson.build:1]: `wihtout` -> `without`

**Review by Conan-Kudo — CHANGES_REQUESTED** [review]

File: test cases/unit/101 python without pkgconfig/meson.build
Comment (nirbheek) [reply test cases/unit/101 python without pkgconfig/meson.build]: This was already merged into master at the time of this comment.

**Comment by nirbheek** [comment]
I've rebased this on top of master and updated the PR title to reflect the remaining issue. @xclaesse do you want to update this PR and get it in for 0.60.2?

**Comment by nirbheek** [comment]
Removed from 0.60.2 milestone.

**Comment by tristan957** [comment]
Anything holding this back guys?

**Comment by xclaesse** [comment]
> Anything holding this back guys?

Rebased in case someone wants to approve it.

**Review by tristan957 — APPROVED** [review]

**Comment by eli-schwartz** [comment]
A year and a half ago this was largely ad-hoc code for finding the dependency. After a lot of refactoring including my uplifting the code to mesonbuild.dependencies.* this actually behaves normally when used as `dependency('python3')`.

We should just use `dep.log_tried()`.

**Comment by xclaesse** [comment]
Ah, no I see, the trick is passing `installation` param to `python_factory()`

**Comment by xclaesse** [comment]
@eli-schwartz Updated to use `detect.find_external_dependency()`. A bit hackish but it makes consistent logs. I hesitated to instead add the factory into `packages` dict with a special key, but that felt worse.

**Comment by eli-schwartz** [comment]
Aha, that turned out to be surprisingly easy. :)

The commit message could do with a bit of updating now, though.

**Comment by xclaesse** [comment]
> The commit message could do with a bit of updating now, though.

updated

**Review by dcbaker — APPROVED** [review]
Ah, yeah. I see how that happened. I'm going to blame this on "dependencies used to be really complicated, now they're only kinda complicated" :)

**Comment by dcbaker** [comment]
but you since @eli-schwartz has been looking at this, probably should have at least an ack from him too.

I do like that you're using `detect.find_external_dependency`. It seems you and Eli have finally made my dream of killing the remains of the old multi-dependency system come true :)

**Comment by eli-schwartz** [comment]
Yes, this is excellent.

**Pull request merged**
